src/markdown/span/emphasis.rs

```diff
@@ -0,0 +1,117 @@
+use crate::markdown::span::parse_spans;
+use crate::markdown::Span;
+use crate::markdown::Span::Emphasis;
+use fancy_regex::Regex;
+
+/// Parses any emphasis (italic) markdown tags in the given text.
+///
+/// # Regex
+///
+/// This thing uses a monster regex from the `simple-markdown` Github project
+/// found [here](https://github.com/Khan/simple-markdown/blob/master/src/index.js#L1607):
+///
+/// ```js
+/// match: inlineRegex(
+///     new RegExp(
+///         // only match _s surrounding words.
+///         "^\\b_" +
+///         "((?:__|\\\\[\\s\\S]|[^\\\\_])+?)_" +
+///         "\\b" +
+///         // Or match *s:
+///         "|" +
+///         // Only match *s that are followed by a non-space:
+///         "^\\*(?=\\S)(" +
+///         // Match at least one of:
+///         "(?:" +
+///           //  - `**`: so that bolds inside italics don't close the
+///           //          italics
+///           "\\*\\*|" +
+///           //  - escape sequence: so escaped *s don't close us
+///           "\\\\[\\s\\S]|" +
+///           //  - whitespace: followed by a non-* (we don't
+///           //          want ' *' to close an italics--it might
+///           //          start a list)
+///           "\\s+(?:\\\\[\\s\\S]|[^\\s\\*\\\\]|\\*\\*)|" +
+///           //  - non-whitespace, non-*, non-backslash characters
+///           "[^\\s\\*\\\\]" +
+///         ")+?" +
+///         // followed by a non-space, non-* then *
+///         ")\\*(?!\\*)"
+///       )
+///   )
+/// ```
+pub fn parse_emphasis(text: &str) -> Option<(Span, usize)> {
+    // Slight hack so I don't have to spend any more time
+    // in Regex Hell.
+    if text.starts_with("***") {
+        return None;
+    }
+
+    lazy_static! {
+        static ref EMPHASIS: Regex = Regex::new(r"^\b_((?:__|\\[\s\S]|[^\\_])+?)_\b|^\*(?=\S)((?:\*\*|\\[\s\S]|\s+(?:\\[\s\S]|[^\s\*\\]|\*\*)|[^\s\*\\])+?)\*(?!\*)").unwrap();
+    }
+
+    match EMPHASIS.is_match(text) {
+        Ok(matches) => {
+            if matches {
+                let captures = EMPHASIS
+                    .captures(text)
+                    .expect("error running regex")
+                    .expect("no match found");
```

Comment:
Maybe instead of doing this, and checking if we have matches in the above `if` statement, we could use matching instead. That would make sure that we're not running the regex twice for some reason (once to check for matches, and again to get the captures; I'm not sure how this regex crate works in this respect).